GoogleIoTCore.agent.nut

```diff
@@ -0,0 +1,1004 @@
+
+                tokenReadyCb({
+                    "jwtToken" : (header + "." + body + "." + signature),
+                    "jwtExpiresAt" : expTime});
+
+            }.bindenv(this)
+        );
+    }
+
+    // Remove the armor, concatenate the lines, and base64 decode the text.
+    function _decodePem(str) {
+        local lines = split(str, "\n");
+        // We really ought to iterate over the array until we find a starting line,
+        // and then look for the matching ending line.
+        if ((lines[0] == "-----BEGIN PRIVATE KEY-----"
+                && lines[lines.len() - 1] == "-----END PRIVATE KEY-----") ||
+            (lines[0] == "-----BEGIN RSA PRIVATE KEY-----"
+                && lines[lines.len() - 1] == "-----END RSA PRIVATE KEY-----") ||
+            (lines[0] == "-----BEGIN PUBLIC KEY-----"
+                && lines[lines.len() - 1] == "-----END PUBLIC KEY-----"))
+        {
+            local all = lines.slice(1, lines.len() - 1).reduce(@(a, b) a + b);
+            return http.base64decode(all);
+        }
+        return null;
+    }
+
+    function _typeof() {
+        return "GoogleIoTCore.Client";
+    }
+}
+
+class GoogleIoTCore.AbstractTransport extends GoogleIoTCore {
+
+    _options            = null;
+
+    _client             = null;
+
+    // Function. Should return a table with jwtToken and expiresAt fields
+    _makeToken          = null;
+
+    _jwtToken           = null;
+    _jwtExpiresAt       = null;
+
+    function _connect() {
+        throw "This method is not implemented";
+    }
+
+    function _disconnect() {
+        throw "This method is not implemented";
+    }
+
+    function _isConnected() {
+        throw "This method is not implemented";
+    }
+
+    function _publish(data, subfolder, onPublished) {
+        throw "This method is not implemented";
+    }
+
+    function _enableCfgReceiving(onReceive, onDone) {
+        throw "This method is not implemented";
+    }
+
+    function _reportState(state, onReported) {
+        throw "This method is not implemented";
+    }
+
+    function _setOnConnected(callback) {
+        throw "This method is not implemented";
+    }
+
+    function _setOnDisconnected(callback) {
+        throw "This method is not implemented";
+    }
+
+    function _setDebug(value) {
+        _debugEnabled = value;
+    }
+
+    function _setClient(client) {
+        _client = client;
+    }
+
+    function _setTokenMaker(maker) {
+        _makeToken = maker;
+    }
+
+    function _updateToken(callback) {
+        if (_makeToken == null) {
+            throw "Token maker is not set";
+        }
+
+        local tokenMade = function (token) {
+            _jwtToken = token.jwtToken;
+            _jwtExpiresAt = token.jwtExpiresAt;
+            callback();
+        }.bindenv(this);
+        _makeToken(tokenMade);
+    }
+}
+
+class GoogleIoTCore.MqttTransport extends GoogleIoTCore.AbstractTransport {
+
+    _stateDisconnected      = true;
```

Comment:
Do we really need `_stateConnected` and `_stateConnected`? The same question for `-ing`.

Reply:
I'm actually thinking of just a single variable `_state` and a enum `STATE`. If several states need to be combined, it can be a bitfield.